### PR TITLE
Support grad backprop when add/sub use broadcast

### PR DESCRIPTION
**Pull request by nogginly**
@christopherzimmerman, this fixes grad backpropagation for addition/subtraction when broadcast occurs because one operand has a different rank than the other. 

* Modified Add/Sub gates to sub-class the `TwoOpGate`
* Made `TwoOpGate` an abstract class with `#backward` abstract method (since every subclass has to implement this method)
* Modified operator +/- definition to pass in the operands 
* Modified add/sub backward methods to accept operands
* Add a private convenience method to backprop the add/sub gradient 
  * I originally had a faster case for handling scalar add/sum, but had to remove it until a future time when `Tensor#sum` has OCL implementation. I'm not sure 
  * Let me know if there's a faster way to implement the rank-by-rank sum that I'm using iteratively.

I've included tests, which are based on results I get from PyTorch.

### CODE REVIEW

**Comment by christopherzimmerman** [comment]
@nogginly thanks for the PR, going to need a bit longer to review this since when I implemented this I guess I made some different design decisions around broadcasted operations than other libraries.

I always set the gradient to match the value that was actually used in the calculation, so in this case, since the broadcast happens during the operation, the gradient will match that shape.  Are you saying that Pytorch aggregates that back down to match the dimensions of the initial variable, before the operation?

**Comment by nogginly** [comment]
> I always set the gradient to match the value that was actually used in the calculation, so in this case, since the broadcast happens during the operation, the gradient will match that shape. Are you saying that Pytorch aggregates that back down to match the dimensions of the initial variable, before the operation?

Hola @christopherzimmerman. Yes, that is correct, I wrote a simple test using PyTorch and got exactly that. I ran into this as I was implementing a two-layer MLP and when I tried to update the the biases using the gradient, the shapes were off and matmul didn't work and that is how I discovered this. The tests I put in are based on those PyTorch test cases.
